engine/common/insight_log/resource_insight_logs.py

```diff
@@ -0,0 +1,132 @@
+        org_id_to_enable_insight_logs, _ = DynamicSetting.objects.get_or_create(
+            name="org_id_to_enable_insight_logs",
+            defaults={"json_value": []},
+        )
+        log_all = "all" in org_id_to_enable_insight_logs.json_value
+        insight_logs_enabled = organization.id in org_id_to_enable_insight_logs.json_value
+        if insight_logs_enabled or log_all:
+            tenant_id = organization.stack_id
+            author_id = author.public_primary_key
+            author = json.dumps(author.username)
+            entity_type = instance.insight_logs_type_verbal
+            try:
+                entity_id = instance.public_primary_key
+            except AttributeError:
+                entity_id = instance.id  # Fallback for entities which have no public_primary_key
+            entity_name = json.dumps(instance.insight_logs_verbal)
+            metadata = instance.insight_logs_metadata
+            log_line = f"tenant_id={tenant_id} author_id={author_id} author={author} action_type=resource action={event.value} resource_type={entity_type} resource_id={entity_id} resource_name={entity_name}"  # noqa
+            for k, v in metadata.items():
+                log_line += f" {k}={json.dumps(v)}"
+            if prev_state and new_state:
+                prev_state, new_state = state_diff_finder(prev_state, new_state)
+                prev_state = escape_json_str_for_insight_log(json.dumps(format_state_for_insight_log(prev_state)))
+                new_state = escape_json_str_for_insight_log(json.dumps(format_state_for_insight_log(new_state)))
+                log_line += f' prev_state="{prev_state}"'
+                log_line += f' new_state="{new_state}"'
+            insight_logger.info(log_line)
+    except Exception as e:
+        logger.warning(f"insight_log.failed_to_write_entity_insight_log exception={e}")
+
+
+def state_diff_finder(before: dict, after: dict):
+    """
+    state_diff_finder finds diff between two serialized representations of the resource
+    """
+    before_diff = {}
+    after_diff = {}
+    for k, v in before.items():
+        if k not in after:
+            before_diff[k] = v
+            continue
+        if after[k] != v:
+            before_diff[k] = before[k]
+            after_diff[k] = after[k]
+    for k, v in after.items():
+        if k not in before:
+            after_diff[k] = v
+    return before_diff, after_diff
+
+
+def escape_json_str_for_insight_log(string):
+    """
+    escape_json_str escapes double quotes near keys and values in json string.
+    it's needed for
+    """
+    return re.sub(r"(?<!\\)(\")", r"\\\1", string)
+
+
+def format_state_for_insight_log(diff_dict):
+    """
+    format_state_for_insight_log formats serialized resource data for the insight log.
+    It hides and prunes fields which shouldn't be exposed
+    """
+    fields_to_prune = ()
+    fields_to_hide = ("verified_phone_number", "unverified_phone_number")
```

Comment:
may be better use whitelist to ensure we don't expose sensitive information?

Reply:
whitelist will be too big,  all fields in almost all models